### PR TITLE
Avoid calling writeStackCookie to early in worker.js

### PR DESCRIPTION
**Pull request by sbc100**
When running a worker we don't want to write the stack cookie until the
worker is actually running a pthread.  This is done in
`establishStackSpace`.

For this reason we don't want to run `writeStackCookie` during
`run` (which is used only to initiialize the worker, before its
running a thread).

The extra call to `writeStackCookie` was not harmfull but its writing
the default stack location (the one used by the main thread!).

### CODE REVIEW

**Review by kripken — COMMENTED** [review]
IIUC pthreads will now only call `writeStackCookie()` - which they always have - but no longer call `stackCheckInit()` - which they used to before this PR, but that was wrong. But do we not need the other parts of `stackCheckInit()`, namely `emscripten_stack_set_limits  /emscripten_stack_init`?

**Comment by sbc100** [comment]
> IIUC pthreads will now only call `writeStackCookie()` - which they always have - but no longer call `stackCheckInit()` - which they used to before this PR, but that was wrong. But do we not need the other parts of `stackCheckInit()`, namely `emscripten_stack_set_limits /emscripten_stack_init`?

I think the answer is that pthreads do all this stuff in `establishStackSpace`: https://github.com/emscripten-core/emscripten/blob/1f6b136d75c9a6a89d07ed069b9cede80b39cebd/src/library_pthread.js#L945-L980

`emscripten_stack_set_limits` and `emscripten_stack_init` both do the same job so only one of them should be called.

In the case of a pthread its the explicit `emscripten_stack_set_limits` that is called from `establishStackSpace`.

File: src/library_dylink.js [hunk 1]
```diff
@@ -646,7 +651,11 @@ var LibraryDylink = {
     return loadModule();
   },
 
-#if STACK_OVERFLOW_CHECK >= 2
+#if STACK_OVERFLOW_CHECK >= 2 && USE_PTHREADS
+  // With USE_PTHREADS we load libraries before we are running a pthread and
+  // therefore before we have a stack.  Instead we calling `__set_stack_limits`
```
Comment (kripken) [inline src/library_dylink.js:656]: I think `calling` is a typo, but I'm not sure what it should be - who calls what?

File: src/postamble.js [hunk 2]
```diff
@@ -214,12 +214,16 @@ function stackCheckInit() {
   // This is normally called automatically during __wasm_call_ctors but need to
   // get these values before even running any of the ctors so we call it redundantly
   // here.
-  // TODO(sbc): Move writeStackCookie to native to to avoid this.
+#if ASSERTIONS && USE_PTHREADS
+  // See $establishStackSpace for the equivelent code that runs on a thread
```
Comment (kripken) [inline src/postamble.js:218]:
```suggestion
  // See $establishStackSpace for the equivalent code that runs on a thread
```

File: src/library_dylink.js [hunk 3]
```diff
@@ -646,7 +651,12 @@ var LibraryDylink = {
     return loadModule();
   },
 
-#if STACK_OVERFLOW_CHECK >= 2
+#if STACK_OVERFLOW_CHECK >= 2 && USE_PTHREADS
+  // With USE_PTHREADS we load libraries before we are running a pthread and
+  // therefore before we have a stack.  Instead we delay calling
+  // `__set_stack_limits` until we start running a thread.  We also need to call
+  // this again for each new thread that the runs on a worker (since each thread
+  // has is own sperate stack region).
```
Comment (kripken) [inline src/library_dylink.js:659]:
```suggestion
  // has its own seperate stack region).
```

**Review by kripken — APPROVED** [review]

**Pull request merged**